vowpalwabbit/core/src/reductions/cb/cb_explore_adf_large_action_space.cc

```diff
@@ -0,0 +1,195 @@
+      : _weights(weights), _column_index(column_index), _seed(seed)
+  {
+  }
+  float operator[](uint64_t index) const
+  {
+    auto combined_index = index + _column_index + _seed;
+    return _weights[index] * merand48_boxmuller(combined_index);
+  }
+};
+
+void cb_explore_adf_large_action_space::predict(VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  predict_or_learn_impl<false>(base, examples);
+}
+void cb_explore_adf_large_action_space::learn(VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  predict_or_learn_impl<true>(base, examples);
+}
+
+cb_explore_adf_large_action_space::cb_explore_adf_large_action_space(uint64_t d, float gamma, VW::workspace* all)
+    : _d(d), _gamma(gamma), _all(all), _seed(all->get_random_state()->get_current_state())
+{
+}
+
+void cb_explore_adf_large_action_space::calculate_shrink_factor(const ACTION_SCORE::action_scores& preds, float min_ck)
+{
+  shrink_factors.clear();
+  for (size_t i = 0; i < preds.size(); i++)
+  { shrink_factors.push_back(std::sqrt(1 + _d + (_gamma / 4.0f * _d) * (preds[i].score - min_ck))); }
+}
+
+inline void just_add_weights(float& p, float, float fw) { p += fw; }
+
+void cb_explore_adf_large_action_space::generate_Q(const multi_ex& examples)
+{
+  // create Q matrix with dimenstions Kxd where K = examples.size()
+  uint64_t num_actions = examples[0]->pred.a_s.size();
+  Q.resize(num_actions, _d);
+
+  // TODO extend wildspace interactions before calling foreach
+  uint64_t row_index = 0;
+  for (auto* ex : examples)
+  {
+    assert(!CB::ec_is_example_header(*ex));
+
+    for (size_t col = 0; col < _d; col++)
+    {
+      float dot_product = 0.f;
+      if (_all->weights.sparse)
+      {
+        LazyGaussianDotProduct<sparse_parameters> w(_all->weights.sparse_weights, col, _seed);
+        GD::foreach_feature<float, float, just_add_weights, LazyGaussianDotProduct<sparse_parameters>>(w,
+            _all->ignore_some_linear, _all->ignore_linear, _all->interactions, _all->extent_interactions,
+            _all->permutations, *ex, dot_product, _all->_generate_interactions_object_cache);
+      }
+      else
+      {
+        LazyGaussianDotProduct<dense_parameters> w(_all->weights.dense_weights, col, _seed);
+        GD::foreach_feature<float, float, just_add_weights, LazyGaussianDotProduct<dense_parameters>>(w,
+            _all->ignore_some_linear, _all->ignore_linear, _all->interactions, _all->extent_interactions,
+            _all->permutations, *ex, dot_product, _all->_generate_interactions_object_cache);
+      }
+
+      Q(row_index, col) = dot_product;
+    }
+    row_index++;
+  }
+}
+
+template <bool is_learn>
+void cb_explore_adf_large_action_space::predict_or_learn_impl(VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  // Explore uniform random an epsilon fraction of the time.
+  if (is_learn) { base.learn(examples); }
+  else
+  {
+    base.predict(examples);
+
```

Comment:
there should be a short circuit if there are less actions than the exploration budget.

Reply:
synced offline, when d > num_actions, the spanner is simply [1:num_actions] (and no need to call SVD)